File: bindings/go/pls4all/sklearn.go
// SPDX-License-Identifier: CeCILL-2.1
//
// Idiomatic Go tier-2 wrapper around pls4all.Fit. Mirrors the
// gonum/learn-style estimator surface: Fit / Predict / Score on a
// receiver struct.
//
// Example:
//
//	m := pls4all.NewPLSRegression(5)  // n_components
//	if err := m.Fit(X, y, n, p); err != nil { ... }
//	preds, err := m.Predict(Xnew, len(Xnew)/p, p)
//	r2 := m.Score(Xtest, ytest, ntest, p)

package pls4all

import (
	"errors"
	"fmt"
	"math"
)

// PLSRegression holds the fitted PLS coefficients + preprocessing
// stats, with a Fit/Predict/Score surface matching the gonum / learn
// idiom.
type PLSRegression struct {
	NComponents int

	// Fitted state (populated by Fit).
	Coefficients []float64 // row-major (p × q)
	XMean        []float64 // (p,)
	YMean        []float64 // (q,)
	NFeaturesIn  int
	NTargets     int
}

// NewPLSRegression returns an unfitted PLS regression with the given
// component count.
func NewPLSRegression(nComponents int) *PLSRegression {
	return &PLSRegression{NComponents: nComponents}
}

// Fit trains the model in-place on (X, y). X is row-major (n × p),
// y is row-major (n × q) or a length-n vector when q == 1.
func (m *PLSRegression) Fit(X, y []float64, n, p int) error {
	if n <= 0 || p <= 0 {
		return errors.New("Fit: n and p must be positive")
	}
	if len(X) != n*p {
		return fmt.Errorf("Fit: len(X)=%d must equal n*p=%d", len(X), n*p)
	}
	if len(y) == 0 || len(y)%n != 0 {
		return fmt.Errorf("Fit: len(y)=%d must be a positive multiple of n=%d",
			len(y), n)
	}
	q := len(y) / n
	res, err := Fit(X, y, n, p, q, m.NComponents)
	if err != nil {
		return err
	}
	m.Coefficients = append([]float64(nil), res.Coefficients...)
	m.XMean = append([]float64(nil), res.XMean...)
	m.YMean = append([]float64(nil), res.YMean...)
	m.NFeaturesIn = p
	m.NTargets = q
	return nil
}

// Predict returns predictions for X (row-major, n_pred × p).
// Following the C ABI convention (cpp/src/core/model.cpp::
// fill_prediction): Y_pred = (X - x_mean) @ coef + y_mean.
func (m *PLSRegression) Predict(X []float64, nPred, p int) ([]float64, error) {
	if m.Coefficients == nil {
		return nil, errors.New("Predict: model is not fitted")
	}
	if p != m.NFeaturesIn {
		return nil, errors.New(
			"Predict: feature count mismatch with fitted model")
	}
	preds := make([]float64, nPred*m.NTargets)
	q := m.NTargets
	for i := 0; i < nPred; i++ {
		for t := 0; t < q; t++ {
			sum := m.YMean[t]
			for j := 0; j < p; j++ {
				centered := X[i*p+j] - m.XMean[j]
				sum += centered * m.Coefficients[j*q+t]
			}
			preds[i*q+t] = sum
		}
	}
	return preds, nil
}

// Score returns the R² coefficient of determination on (X, y).
func (m *PLSRegression) Score(X, y []float64, n, p int) (float64, error) {
	preds, err := m.Predict(X, n, p)
	if err != nil {
		return 0, err
	}
	q := m.NTargets
	// uniform_average across targets.
	var r2sum float64
	for t := 0; t < q; t++ {
		var mean float64
		for i := 0; i < n; i++ {
			mean += y[i*q+t]
		}
		mean /= float64(n)
		var ssRes, ssTot float64
		for i := 0; i < n; i++ {
			dr := preds[i*q+t] - y[i*q+t]
			dt := y[i*q+t] - mean
			ssRes += dr * dr
			ssTot += dt * dt
		}
		if ssTot == 0 || math.IsNaN(ssTot) {
			continue
		}
		r2sum += 1 - ssRes/ssTot
	}
	return r2sum / float64(q), nil
}

File: bindings/_archive/rust/pls4all/src/sklearn.rs
// SPDX-License-Identifier: CECILL-2.1
//
// Idiomatic Rust tier-2 wrapper around `pls4all::pls_fit`. Provides a
// `PLSRegression` struct with a `fit` / `predict` / `score` surface
// consistent with the `linfa` ecosystem's `Fit` / `Predict` traits,
// without taking a hard dependency on linfa itself.
//
// Example:
//
// ```ignore
// use pls4all::sklearn::PLSRegression;
// let model = PLSRegression::new(5).fit(&x, &y, n, p)?;
// let preds = model.predict(&x_new, n_new, p)?;
// let r2 = model.score(&x_test, &y_test, n_test, p)?;
// ```

use crate::{pls_fit, FitError};

#[derive(Debug, Clone)]
pub struct PLSRegression {
    pub n_components: usize,

    coefficients: Option<Vec<f64>>,
    x_mean: Option<Vec<f64>>,
    y_mean: Option<Vec<f64>>,
    n_features_in: usize,
    n_targets: usize,
}

impl PLSRegression {
    /// Create an unfitted PLS regression with the given component count.
    pub fn new(n_components: usize) -> Self {
        PLSRegression {
            n_components,
            coefficients: None,
            x_mean: None,
            y_mean: None,
            n_features_in: 0,
            n_targets: 0,
        }
    }

    /// Train the model in-place. `x` is row-major `(n, p)`; `y` is
    /// row-major `(n, q)` or a length-`n` vector when `q == 1`.
    pub fn fit(mut self, x: &[f64], y: &[f64], n: usize,
                p: usize) -> Result<Self, FitError> {
        if n == 0 || p == 0 {
            return Err(FitError::NComponents(0));
        }
        if x.len() != n * p {
            return Err(FitError::XLength { got: x.len(), expected: n * p });
        }
        if y.is_empty() || y.len() % n != 0 {
            return Err(FitError::YLength { got: y.len(), expected: n });
        }
        let q = y.len() / n;
        let res = pls_fit(x, y, n, p, q, self.n_components)?;
        self.coefficients = Some(res.coefficients);
        self.x_mean = Some(res.x_mean);
        self.y_mean = Some(res.y_mean);
        self.n_features_in = p;
        self.n_targets = q;
        Ok(self)
    }

    /// Predict on new X (row-major `(n_pred, p)`).
    pub fn predict(&self, x: &[f64], n_pred: usize,
                    p: usize) -> Result<Vec<f64>, FitError> {
        let coef = self.coefficients.as_ref().ok_or(FitError::Native(-1))?;
        let xmean = self.x_mean.as_ref().unwrap();
        let ymean = self.y_mean.as_ref().unwrap();
        if p != self.n_features_in {
            return Err(FitError::Native(-2));
        }
        let q = self.n_targets;
        let mut preds = vec![0.0_f64; n_pred * q];
        for i in 0..n_pred {
            for t in 0..q {
                let mut sum = ymean[t];
                for j in 0..p {
                    let centered = x[i * p + j] - xmean[j];
                    sum += centered * coef[j * q + t];
                }
                preds[i * q + t] = sum;
            }
        }
        Ok(preds)
    }

    /// R² coefficient of determination on (X, y), uniform-averaged
    /// across targets.
    pub fn score(&self, x: &[f64], y: &[f64], n: usize,
                  p: usize) -> Result<f64, FitError> {
        let preds = self.predict(x, n, p)?;
        let q = self.n_targets;
        let mut r2_sum = 0.0_f64;
        for t in 0..q {
            let mean: f64 = (0..n).map(|i| y[i * q + t]).sum::<f64>()
                / n as f64;
            let mut ss_res = 0.0_f64;
            let mut ss_tot = 0.0_f64;
            for i in 0..n {
                let dr = preds[i * q + t] - y[i * q + t];
                let dt = y[i * q + t] - mean;
                ss_res += dr * dr;
                ss_tot += dt * dt;
            }
            if ss_tot != 0.0 && ss_tot.is_finite() {
                r2_sum += 1.0 - ss_res / ss_tot;
            }
        }
        Ok(r2_sum / q as f64)
    }

    pub fn coefficients(&self) -> Option<&[f64]> {
        self.coefficients.as_deref()
    }
    pub fn x_mean(&self) -> Option<&[f64]> {
        self.x_mean.as_deref()
    }
    pub fn y_mean(&self) -> Option<&[f64]> {
        self.y_mean.as_deref()
    }
}
